### ISSUE TITLE
Zip Operator Breaks with 7 or more Observables with non-primitive types

### ISSUE THREAD

**Issue by bd60** [ReactiveX/rxjs#2782]
**RxJS version:**
5.1.0
**Code to reproduce:**

      var zipped$ = Observable.zip(
             Observable.of({prop1:1}),
             Observable.of({prop2:2}),
             Observable.of({prop3:3}),
             Observable.of({prop4:4}),
             Observable.of({prop5:5}),
             Observable.of({prop6:6}),
             Observable.of({prop7:7}),
         );
        var reduced$ = zipped$.mergeAll().reduce((acc,val) => Object.assign({}, acc, val));

gives error:  "property 'reduce' does not exist on type '{}'"

However, this works:

     var zipped$ = Observable.zip(
            Observable.of({prop1:1}),
            Observable.of({prop2:2}),
            Observable.of({prop3:3}),
            Observable.of({prop4:4}),
            Observable.of({prop5:5}),
            Observable.of({prop6:6}),
        );

        var reduced$ = zipped$.mergeAll().reduce((acc,val) => Object.assign({}, acc, val));

this works: 

     var zipped$ = Observable.zip(
            Observable.of(1),
            Observable.of(2),
            Observable.of(3),
            Observable.of(4),
            Observable.of(5),
            Observable.of(6),
            Observable.of(7),
        );

        var reduced$ = zipped$.mergeAll().reduce((acc,val) => Object.assign({}, acc, val));

And this works:

     var zipped$ = Observable.zip(
            Observable.of({prop1:1}),
            Observable.of({prop2:2}),
            Observable.of({prop3:3}),
            Observable.of({prop4:4}),
            Observable.of({prop5:5}),
            Observable.of({prop6:6}),
            Observable.of({prop7:7}),
            function(...args: any[]) {
               return args;
            }
        );

        var reduced$ = zipped$.mergeAll().reduce((acc,val) => Object.assign({}, acc, val));

**Expected behavior:**

It should work with more than 6 non primitive arguments, or the behavior should at least be consistent or documented.

**Actual behavior:**

Noted above but, gives error:  "property 'reduce' does not exist on type '{}'"

**Additional information:**

**Comment by kwonoj** [comment]
If you check out definition of operator (https://github.com/ReactiveX/rxjs/blob/master/src/operator/zip.ts#L13-L26) not only zip, other operators like `combinLatest` as well - it is all manually handcrafted. 

Unfortunately there isn't great way to automatically create overloaded signature to support multiple paramaters, me, @david-driscoll and other core members discussed and settled down with some specific number (6, currently) cause we thought there won't be frequent use cases more than those numbers.

In short, it's known / and our compromise with limitation. If there's graceful way to support this, it'd be awesome.

**Comment by kwonoj** [comment]
I'll leave this opened for few more days if things has been changes with latest compiler, and if @david-driscoll can chime in bit more.

**Comment by bd60** [comment]
Thanks for the explanation.  At least I understand why this is happening and it makes sense. I would just suggest that the limitation be added to the documentation.

The use case for so many arguments, in case you're interested, is a form container component that has a lot of sub forms in it.  There is a comm service that as a saveAll action on it that causes each component to call it's own save method through the comm service.

it looks like this:

    var allSaved$ = Observablezip(
        this.commSvc.componentOneSaved$,
        this.commSvc.componentTwoSaved$,
        this.commSvc.componentThreeSaved$,
        this.commSvc.componentFourSaved$,
        this.commSvc.componentFiveSaved$,
        this.commSvc.componentSixSaved$,
        this.commSvc.componentSevenSaved$)
    .take(1).mergeAll().reduce((acc,val) => Object.assign({},acc,val));

    this.commSvc.onSaveAll$.switchMap(e => allSaved$)
        .switchMap(final => this.dataSvc.save(final)).subscribe(e => this.saveSuccess());


The container component needs the data from all of the sub components to save, and it needs exactly one response from all of them after the saveAll observable is triggered.  If there's a better means of accomplishing this, I'd love to know about it.  Like some kind of operator or subject variation that collects a response from all subscribers, which is easily done with a hand written observer pattern.

**Comment by kwonoj** [comment]
Most simpliest workaround for zip at this moment is just supplying array or rest array, which doesn't hit overload cases and falls back signature into default one. i.e

```
const a = [Rx.Observable.of(1), ....]

Rx.Observable.zip(a); 
//or
Rx.Observable.zip(...a);
```

**Comment by bd60** [comment]
Thank you! much cleaner than supplying the projection function

**Comment by kwonoj** [comment]
I'm going to close this for now.
